# Машинное обучение, ФКН ВШЭ

# Практическое задание 11. Поиск ближайших соседей

## Общая информация

Дата выдачи: 05.04.2021

Мягкий дедлайн: 14.04.2021 00:59 MSK

Жёсткий дедлайн: 18.04.2021 00:59 MSK

## Оценивание и штрафы

Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 6.51 балла.



Сдавать задание после указанного жёсткого срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

Неэффективная реализация кода может негативно отразиться на оценке.

## Формат сдачи

Задания сдаются через систему anytask. Посылка должна содержать:

* Ноутбук homework-practice-11-knn-Username.ipynb

Username — ваша фамилия и имя на латинице именно в таком порядке.

In [2]:

import matplotlib.pyplot as plt
import numpy as np
import os
import random

from tqdm.notebook import tqdm

Возьмем [датасет](https://www.kaggle.com/delayedkarma/impressionist-classifier-data)  с картинами известных импрессионистов. Работать будем не с самими картинками, а с эмбеддингами картинок, полученных с помощью сверточного классификатора.

![](https://storage.googleapis.com/kagglesdsdata/datasets/568245/1031162/training/training/Gauguin/190448.jpg?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=databundle-worker-v2%40kaggle-161607.iam.gserviceaccount.com%2F20210405%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210405T125358Z&X-Goog-Expires=172799&X-Goog-SignedHeaders=host&X-Goog-Signature=a271b474bf9ec20ba159b951e0ae680fc2b0c694666031f7ea6fc39598172cc55e10f75c12b678b21da9e6bdc20e46886133c219625648b407d2f600eebfdda909b29e0f7f13276d8fea2f8d0480d6298bd98e7f118eb78e8b632fc3d141365356b0e3a2fdd4f09119f99f0907a31da62e8dae7e625e32d831238ecc227b1f5ad2e96a8bfb43d93ef6fe88d7e663e51d387d3550dcad2a7eefc5c941028ba0d7751d18690cf2e26fcdfaa4dacd3dcbb3a4cbb355e62c08b158007b5e764e468cecd3292dae4cfc408e848ecf3e0e5dbe5faa76fcdd77d5370c868583c06e4e3d40c73a7435bd8c32a9803fe6b536e1c6f0791219aadd06120291e937e57c214a)

In [3]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

In [4]:
X_train = np.load('embeddings/embeds_avpool_train.npy')
y_train = np.load('embeddings/labels_avpool_train.npy')
X_test = np.load('embeddings/embeds_avpool_test.npy')
y_test = np.load('embeddings/labels_avpool_test.npy')

Будем смотреть на обычную долю верных ответов и на долю верных ответов в топ-3.

In [5]:
def top_3_accuracy_score(y_true, probas):
    preds = np.argsort(probas, axis=1)[:, -3:]
    matches = np.zeros_like(y_true)
    for i in range(3):
        matches += (preds[:, i] == y_true)
    return matches.sum() / matches.size

def scorer(estimator, X, y):
    return accuracy_score(y, estimator.predict(X))

**Задание 1. (1 балл)**

Обучите классификатор k ближайших соседей (из sklearn) на данных, подобрав лучшие гиперпараметры. Замерьте качество на обучающей и тестовой выборках.

In [15]:
parameters = {'n_neighbors':range(3, 20, 3), 'p':[1, 2, 3]}
clf = GridSearchCV(KNeighborsClassifier(), parameters, scoring=scorer)
clf.fit(X_train, y_train)
clf.cv_results_

{'mean_fit_time': array([0.18979769, 0.20349851, 0.19203858, 0.19658499, 0.19350042,
        0.19011741, 0.19559083, 0.19474583, 0.19483786, 0.1878469 ,
        0.18801403, 0.19192195, 0.18654523, 0.18605242, 0.20307403,
        0.19946294, 0.19191074, 0.1931777 ]),
 'std_fit_time': array([0.00797855, 0.01423753, 0.00813648, 0.00503222, 0.00615641,
        0.00904779, 0.01036489, 0.00782459, 0.01572994, 0.0085735 ,
        0.00390001, 0.00824591, 0.00592955, 0.00554101, 0.01258571,
        0.00458366, 0.0085507 , 0.00874293]),
 'mean_score_time': array([ 1.15351782,  1.09412475, 12.52161975,  1.1819983 ,  1.1915916 ,
        13.91590638,  1.20401587,  1.22999249, 14.78482475,  1.12162981,
         1.21681685, 15.26096005,  1.16468663,  1.22001686, 15.9761394 ,
         1.17310238,  1.27638078, 16.30191245]),
 'std_score_time': array([0.03588963, 0.02849217, 0.09848884, 0.01589384, 0.03838115,
        0.15176321, 0.08587988, 0.04521832, 0.25109943, 0.01188209,
        0.04003295, 0.2067

In [23]:
np.argmax(clf.cv_results_['mean_test_score']) #n_neighbors = 15, p=1

12

Оптимальным получилось p=1, но по результатам видно, что не очень сильно отличается от резов p=2. Поэтому в дальнейшем оставим p=2, т.к. оно учится вроде быстрее 

In [25]:
knn = KNeighborsClassifier(n_neighbors=15, p=1)
knn.fit(X_train, y_train)
print(accuracy_score(y_train, knn.predict(X_train))) # качество на трейне
print(accuracy_score(y_test, knn.predict(X_test))) # качество на тесте

0.6439317953861585
0.5464646464646464


Видим, что на тесте качество нормальное такое (случайный 0.1 выдает)

**Задание 2. (2 балла)** 

Теперь будем пользоваться метрикой Махалонобиса. Обучите её одним из методов [отсюда](http://contrib.scikit-learn.org/metric-learn/supervised.html). Напомним, что вычисление метрики Махалонобиса эквивалентно вычислению евклидова расстояния между объектами, к которым применено некоторое линейное преобразование (вспомните семинары). Преобразуйте данные и обучите kNN на них, перебрав гиперпараметры, замерьте качество.

Заметим, что в библиотеке metric-learn есть несколько способов обучать матрицу преобразования. Выберите лучший, аргументируйте свой выбор.

Note: Некоторые методы с дефолтными параметрами учатся очень долго, будьте внимательны


In [5]:
# NCA
d = X_train.shape[1] 
from metric_learn import NCA
nca = NCA(n_components=d, max_iter=50)
nca.fit(X_train, y_train)

NCA(max_iter=50, n_components=256)

In [6]:
X_train_new = nca.transform(X_train)

In [7]:
parameters = {'n_neighbors':range(3, 40, 5)}
clf = GridSearchCV(KNeighborsClassifier(), parameters, scoring=scorer)
clf.fit(X_train_new, y_train)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': range(3, 40, 5)},
             scoring=<function scorer at 0x000001D222EB3798>)

In [8]:
clf.best_params_

{'n_neighbors': 13}

In [9]:
knn = KNeighborsClassifier(n_neighbors=13)
knn.fit(X_train_new, y_train)
print(accuracy_score(y_train, knn.predict(X_train_new))) # качество на трейне
print(accuracy_score(y_test, knn.predict(nca.transform(X_test)))) # качество на тесте

0.660481444332999
0.5636363636363636


In [31]:
# LMNN
d = X_train.shape[1] 
from metric_learn import LMNN
lmnn = LMNN(n_components=d, max_iter=50)
lmnn.fit(X_train, y_train)
X_train_new = lmnn.transform(X_train)
parameters = {'n_neighbors':range(3, 40, 5)}
clf = GridSearchCV(KNeighborsClassifier(), parameters, scoring=scorer)
clf.fit(X_train_new, y_train)
knn = KNeighborsClassifier(n_neighbors=clf.best_params_['n_neighbors'])
knn.fit(X_train_new, y_train)
print(accuracy_score(y_train, knn.predict(X_train_new))) # качество на трейне
print(accuracy_score(y_test, knn.predict(lmnn.transform(X_test)))) # качество на тесте

0.6331494483450351
0.5424242424242425


In [29]:
# LFDA # а тут решил еще распределение весов перебрать
d = X_train.shape[1] 
from metric_learn import LFDA
lfda = LFDA()
lfda.fit(X_train, y_train)
X_train_new = lfda.transform(X_train)
parameters = {'n_neighbors':range(10, 80, 10), 'weights':['uniform', 'distance']}
clf = GridSearchCV(KNeighborsClassifier(), parameters, scoring=scorer)
clf.fit(X_train_new, y_train)
knn = KNeighborsClassifier(n_neighbors=clf.best_params_['n_neighbors'], weights=clf.best_params_['weights'])
knn.fit(X_train_new, y_train)
print(accuracy_score(y_train, knn.predict(X_train_new))) # качество на трейне
print(accuracy_score(y_test, knn.predict(lfda.transform(X_test)))) # качество на тесте

C:\Users\Danill\AppData\Local\Programs\Python\Python37\lib\site-packages\scipy\sparse\linalg\eigen\arpack\arpack.py:1592: RuntimeWarning: k >= N for N * N square matrix. Attempting to use scipy.linalg.eigh instead.
  RuntimeWarning)
1.0
0.5444444444444444


In [32]:
# MLKR
d = X_train.shape[1] 
from metric_learn import MLKR
mlkr = MLKR(n_components=d, max_iter=50)
mlkr.fit(X_train, y_train)
X_train_new = mlkr.transform(X_train)
parameters = {'n_neighbors':range(5, 40, 5)}
clf = GridSearchCV(KNeighborsClassifier(), parameters, scoring=scorer)
clf.fit(X_train_new, y_train)
knn = KNeighborsClassifier(n_neighbors=clf.best_params_['n_neighbors'])
knn.fit(X_train_new, y_train)
print(accuracy_score(y_train, knn.predict(X_train_new))) # качество на трейне
print(accuracy_score(y_test, knn.predict(mlkr.transform(X_test)))) # качество на тесте

0.6893179538615848
0.5363636363636364


Видно, что все методы справились примерно одинакого.
качество на тесте 0.55 (у NCA 0.56) примерно у всех. Для 10 классов это вроде круто.

**Задание 3. (1 балл)** 

Что будет, если в качестве матрицы в расстоянии Махалонобиса использовать случайную матрицу? Матрицу ковариаций?

In [9]:
# Случайная матрица 
from sklearn.neighbors import DistanceMetric
A = np.random.rand(d, d)
L = np.linalg.inv(A.T @ A) 
knn = KNeighborsClassifier(n_neighbors=20,
 metric='mahalanobis',
 metric_params={'V':L}, 
 n_jobs=-1)
knn.fit(X_train, y_train)
print(accuracy_score(y_train, knn.predict(X_train))) # качество на трейне
print(accuracy_score(y_test, knn.predict(X_test))) # качество на тесте

C:\Users\Danill\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\neighbors\_base.py:492: RuntimeWarning: invalid value encountered in sqrt
  return tree.query(*args, **kwargs)
0.11634904714142427
C:\Users\Danill\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\neighbors\_base.py:492: RuntimeWarning: invalid value encountered in sqrt
  return tree.query(*args, **kwargs)
0.10505050505050505


In [11]:
# Матрица ковариаций
L = np.linalg.inv(np.cov(X_train, rowvar=False))
#L = np.cov(X_train, rowvar=False)
knn = KNeighborsClassifier(n_neighbors=20,
 metric='mahalanobis',
 metric_params={'V':L}, 
 n_jobs=-1)
knn.fit(X_train, y_train)
print(accuracy_score(y_train, knn.predict(X_train))) # качество на трейне
print(accuracy_score(y_test, knn.predict(X_test))) # качество на тесте

C:\Users\Danill\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\neighbors\_base.py:492: RuntimeWarning: invalid value encountered in sqrt
  return tree.query(*args, **kwargs)
0.1945837512537613
C:\Users\Danill\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\neighbors\_base.py:492: RuntimeWarning: invalid value encountered in sqrt
  return tree.query(*args, **kwargs)
0.20202020202020202


Произошло то, чего стоило ожидать.
Случайная матрица расстояний дала качество случайного классификатора (0.1).
А ковариационная отработала в 2 раза лучше (качество на тесте 0.2), но все еще сильно хуже продвинутых методов типа NCA.

**Задание 4. (1 балл)** Обучите какой-нибудь градиентный бустинг на обычных и трансформированных наборах данных, замерьте качество, задумайтесь о целесообразности других методов.

In [15]:
from catboost import CatBoostClassifier
clf = CatBoostClassifier(verbose=False)
clf.fit(X_train, y_train)
print(accuracy_score(y_train, clf.predict(X_train))) # качество на трейне
print(accuracy_score(y_test, clf.predict(X_test))) # качество на тесте

1.0
0.6151515151515151


In [20]:
nca = NCA(n_components=d, max_iter=50)
clf.fit(nca.fit_transform(X_train, y_train), y_train)
print(accuracy_score(y_train, clf.predict(nca.transform(X_train)))) # качество на трейне
print(accuracy_score(y_test, clf.predict(nca.transform(X_test)))) # качество на тесте

0.9997492477432297
0.6111111111111112


In [21]:
from metric_learn import LFDA
lfda = LFDA()
clf.fit(lfda.fit_transform(X_train, y_train), y_train)
print(accuracy_score(y_train, clf.predict(lfda.transform(X_train)))) # качество на трейне
print(accuracy_score(y_test, clf.predict(lfda.transform(X_test)))) # качество на тесте

C:\Users\Danill\AppData\Local\Programs\Python\Python37\lib\site-packages\scipy\sparse\linalg\eigen\arpack\arpack.py:1592: RuntimeWarning: k >= N for N * N square matrix. Attempting to use scipy.linalg.eigh instead.
  RuntimeWarning)
1.0
0.6272727272727273


Видно, что от преобразования данных не сильно что-то зависит (ну оно и ясно, это же деревья, которые как-то разбивают пространство. От того, что мы это пространство как-то преобразуем, ничего не поменяется). Итог: бустинг на дефолтных настройках и сырых данных всех победил (качество на тесте 0.62). Над целесообразностью задумался. Как придумаю - напишу.

**Бонус. (1 балл)**

Достигните доли верных ответов 0.75 на тестовой выборке, не используя нейросети.

**Бонус. (0.5 балла)**

Если вам понравилось это задание и вы хотите приобщиться к искусству, настоятельно рекомендуем отвлечься от учебы, найти себе компанию и сходить в [Пушкинский музей](https://pushkinmuseum.art/visitors/index.php?lang=ru), в которым традиционно выставляются картины известных импрессионистов 20-го века. В данный момент там проходит временная экспозиция "[Не живопись.](https://pushkinmuseum.art/events/archive/2021/exhibitions/gallery/index.php)", где выставлены работы импрессионистов на керамике и текстиле.

Если компания не находится, а одному идти совсем не хочется, можно обратиться за помощью к ассистентам или преподавателям, это станет отличным поводом наладить контакт с ними.

В качестве ответа на задание необходимо прислать фотоотчет, на котором видно, что вы посетили это замечательное место после выдачи ДЗ.

**Бонус. (0.01 балла)**

Предложите самый лучший каомодзи и объясните свой выбор (в комментариях не нуждается)

Боян, но мб кто не видел 
( . Y . ) 